File: cosmos-master/code/sorting/heap_sort/heap_sort.js
/* Part of Cosmos by OpenGenus Foundation */
  var length;
  /* to create MAX  array */
  function heap_root(array, i) {
      var left = 2 * i + 1;
      var right = 2 * i + 2;
      var max = i;

      if (left < length && array[left] > array[max]) {
          max = left;
      }

      if (right < length && array[right] > array[max]) {
          max = right;
      }

      if (max != i) {
          swap(array, i, max);
          heap_root(array, max);
      }
  }

  function swap(array, index_A, index_B) {
      var temp = array[index_A];

      array[index_A] = array[index_B];
      array[index_B] = temp;
  }

  function heapSort(array) {

      length = array.length;

      for (var i = Math.floor(length / 2); i >= 0; i -= 1) {
          heap_root(array, i);
      }

      for (i = array.length - 1; i > 0; i--) {
          swap(array, 0, i);
          length--;


          heap_root(array, 0);
      }
  }

  // var arr = [3, 0, 2, 5, -1, 4, 1];
  // heapSort(arr);
  // console.log(arr);

File: cosmos-master/code/sorting/heap_sort/heap_sort.py
# Part of Cosmos by OpenGenus Foundation
import math

def heapRoot(arr, i, length):
	left = 2*i +1
	right = 2*i + 2
	max_index = i

	if left<length and arr[left]>arr[max_index]:
		max_index = left

	if right<length and arr[right]>arr[max_index]:
		max_index = right

	if max_index != i:
		swap(arr, max_index, i)
		heapRoot(arr, max_index, length)

def swap(arr, a, b):
	arr[a] = arr[a]+arr[b]
	arr[b] = arr[a]-arr[b]
	arr[a] = arr[a]-arr[b]

#sorting is in the ascending order
def heapSort(arr):
	#form a max heap
	for i in range(math.floor(len(arr)/2), -1, -1):
		heapRoot(arr, i, len(arr))

	length = len(arr)
	#remove the root and reform the heap
	for i in range(len(arr)-1, 0, -1):
		print(i)
		swap(arr, 0, i)
		print(arr)
		length -= 1
		heapRoot(arr, 0, length)

#arr = [7, -1, -2, 5, 15, 0, 10]
#heapSort(arr)
#print(arr)
